yabt/builders/cpp.py

```diff
@@ -21,38 +21,169 @@
+    register_manipulate_target_hook)
 from ..logging import make_logger
-from ..utils import yprint
+from ..target_utils import split
+from ..utils import link_artifacts, yprint
 
 
 logger = make_logger(__name__)
 
 
-register_app_builder_sig('CppApp', [('executable', PT.File)])
+register_app_builder_sig(
+    'CppApp', [('executable', PT.File, None), ('main', PT.Target, None)])
 
 
 @register_manipulate_target_hook('CppApp')
 def cpp_app_manipulate_target(build_context, target):
     logger.debug('Injecting "{}" to deps of {}',
                  target.props.base_image, target)
     target.deps.append(target.props.base_image)
+    if target.props.main and target.props.main not in target.deps:
+        logger.debug('Injecting "{}" to deps of {}',
+                     target.props.main, target)
+        target.deps.append(target.props.main)
 
 
 @register_build_func('CppApp')
-def cpp_bin_builder(build_context, target):
+def cpp_app_builder(build_context, target):
     """Pack a C++ binary as a Docker image with its runtime dependencies.
 
     TODO(itamar): Dynamically analyze the binary and copy shared objects
     from its buildenv image to the runtime image, unless they're installed.
     """
     yprint(build_context.conf, 'Build CppApp', target)
-    if target.props.executable not in target.artifacts['app']:
-        target.artifacts['app'].append(target.props.executable)
+    if target.props.executable and target.props.main:
+        raise KeyError(
+            '`main` and `executable` arguments are mutually exclusive')
+    if target.props.executable:
+        if target.props.executable not in target.artifacts['app']:
+            target.artifacts['app'].append(target.props.executable)
+        entrypoint = [target.props.executable]
+    elif target.props.main:
+        prog = build_context.targets[target.props.main]
+        entrypoint = [join('/usr/src/gen',
+                           list(prog.artifacts['gen'].keys())[0])]
+    else:
+        raise KeyError('Must specify either `main` or `executable` argument')
     build_app_docker_and_bin(
-        build_context, target, entrypoint=[target.props.executable])
+        build_context, target, entrypoint=entrypoint)
+
+
+register_builder_sig(
+    'CppProg',
+    [('binary', PT.File),
+     ('sources', PT.FileList),
+     ('in_buildenv', PT.Target),
+     ('headers', PT.FileList, None),
+     ('cmd_env', None),
+     # ('copy_bin_to', PT.File, None),
+     ])
+
+
+@register_manipulate_target_hook('CppProg')
+def cpp_prog_manipulate_target(build_context, target):
+    target.buildenv = target.props.in_buildenv
+
+
+def compile_cc(build_context, buildenv, sources, workspace_dir,
+               buildenv_workspace, cmd_env):
+    """Compile list of C++ source files in a buildenv image
+       and return list of generated object file.
+    """
+    objects = []
+    for src in sources:
+        obj_rel_path = '{}.o'.format(splitext(src)[0])
```

Comment:
This means we stop coping the sources - right ?

Reply:
copying what sources from where to where?

since compile & link run in sandboxed environments, we hard-link the sources from the project dir to the compile dir, and then hard-link the objects from where they were created to wherever they are needed.

Reply:
My concern was integration of error messages in IDEs; we agreed to put a low priority TODO

Reply:
added TODO